### PR TITLE
Static typing with typescript

### PR DESCRIPTION
**Pull request by jbaxleyiii**
This PR is alternative way to solve #6 using typescript instead of flow.

### Pros:
- typescript seems to be a more robust typings solution. Things that flow allowed, typescript has flagged. (Although my flow setup is probably not 100%)
- typescript auto generates type definitions for source methods. This makes exporting types trivial. (Again, this could be possible in flow but from what I saw, you have to manually define export types in flow)
- singular build + typechecking
- more robust editor integration

### Cons:
- no support for ES7/8 spread operators `let foo = { ...bar }`

### CODE REVIEW

**Comment by stubailo** [comment]
Actually I feel like both things listed as drawbacks for typescript here are actually issues with my code!

1. My argument serialization for the cache key is not complete, and just doesn't handle list values properly, among other things
2. I think it isn't a good idea to change the type of a value mid-function so I think it would be best to change that code!

**Comment by jbaxleyiii** [comment]
@stubailo I've made a branch to verify the CI will handle typescript correctly and highlight the current issues. If we want to slowly adopt typescript, we can use `any` in places where we still aren't strict enough with types? 

here is the branch https://github.com/apollostack/apollo-client/tree/typescript-weak

File: src/parser.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
+/// <reference path="../typings/browser/ambient/es6-promise/index.d.ts" />
+/// <reference path="../typings/browser/ambient/graphql/index.d.ts" />
+/// <reference path="../typings/browser/definitions/lodash/index.d.ts" />
```
Comment (mquandalle) [inline src/parser.ts]: I think you can avoid these `<reference path />` on top of every file by including the `/typings` directory to the list of typescript files in `tsconfig.json`.

**Comment by mquandalle** [comment]
On object spread, I agree that it’s a bit unfortunate that Typescript doesn’t support it given how popular they are in the Redux/functionalJS communities. Hopefully it will come sooner than latter, https://github.com/Microsoft/TypeScript/issues/2103, and so using `Object.assign` for now and replacing them by object spread when it arrives is reasonable.

**Comment by stubailo** [comment]
Hmm, that's kind of a bummer - it's a reminder that TypeScript isn't compatible with all future JavaScript features that Babel might be. Perhaps it is worth it to just avoid those features to get the enhanced tooling...

**Comment by jbaxleyiii** [comment]
So far the spread operator is the only big miss I've seen. I love `...foo` but I do think the tooling out weighs it. Plus the timeline for it being added is within the next month or so :tada:

**Comment by stubailo** [comment]
@jbaxleyiii I'm going to try to merge like all of these PRs today, if that's fine with you?

**Comment by jbaxleyiii** [comment]
@stubailo that would be great! I'd recommend merging #11 instead of this one though since it will let us adopt fully typed more slowly

**Comment by stubailo** [comment]
@jbaxleyiii after some in-depth conversation with @helfer I feel like it might be better to adopt Flow instead of TypeScript after all for the following reasons (assuming they both support basic typing requirements):

1. It will be nicer for potential users of this library that it's a codebase built using Babel that they can understand, with `.js` file extensions. I think Flow looks a lot more like "annotations on top of JavaScript" vs. TypeScript which is more like "a new programming language that looks like JavaScript".
2. People looking to use Apollo will have to accept using tools developed by Facebook, specifically GraphQL. It might be easier to convince those people that Flow is a reasonable choice than also having to trust Microsoft.
3. After looking into TypeScript a bit, it looks like it doesn't do well in cases where not all of your stuff is typed - it's a bit harder to incrementally adopt (as you've seen in these two PRs)

The next step for me is, I'm going to take your Flow PR and try to set up the right tooling in Atom to get autocompletion, etc. If it works well I think I'm going to call it in favor of Flow. If it's too hard to set up I'm going to do the same with TypeScript to see if it's significantly better.

Does this sound like a reasonable choice?

**Comment by jbaxleyiii** [comment]
@stubailo that sounds good to me. I like typescript but flow is way easier to get into!

**Comment by stubailo** [comment]
I'm starting to go back on this, after trying to get Flow to work... have you actually been able to get it to report a useful type error? Like it doesn't actually seem to check if the fields we are using are on the GraphQL AST...?

**Comment by mquandalle** [comment]
On @stubailo point `3.`, maybe I’m underestimating the amount of development that went into the Apollo client, but is the difficulty to incremental adopt typing is such a big deal? Mmaybe it’s worth spending a few days to have 100% type coverage (whichever type system end up being chosen)?

**Comment by stubailo** [comment]
>  is the difficulty to incremental adopt typing is such a big deal?

I agree! I'm just wondering about stuff like NPM packages and the like. In TypeScript it seems like we basically need to have 100% coverage on these? I could be misinterpreting though.

**Comment by mquandalle** [comment]
Haven’t tried it yet but Typescript 1.8 introduced the ability to mix JavaScript and Typescript in the same application ([see the release notes](https://github.com/Microsoft/TypeScript/wiki/Roadmap#18)). (and there are already many libraries on DefinitelyTyped)

**Comment by stubailo** [comment]
> Can't change type of var on the fly

The way you do it is by defining a new var:

```js
const myVar: string | number;
const myNumberVar =

**Comment by jbaxleyiii** [comment]
@stubailo you fixed all the typing errors :tada: :tada:

**Comment by stubailo** [comment]
Yeah arguably a lot of them were actual flaws in the code!

**Comment by stubailo** [comment]
I'm really loving VS Code btw. This is super dope.

**Comment by jbaxleyiii** [comment]
@stubailo the auto type generation is really great as a consumable:

```
import { Document } from 'graphql';
export declare function readQueryFromStore({store, query}: {
    store: Object;
    query: Document | string;
}): Object;
export declare function readFragmentFromStore({store, fragment, rootId}: {
    store: Object;
    fragment: Document | string;
    rootId: string;
}): Object;
```

**Comment by stubailo** [comment]
@jbaxleyiii so should I merge this branch now, or what? Also, are you working on this full-time? If so, we should get on some kind of chat channel or something.

**Comment by jbaxleyiii** [comment]
@stubailo I'm all for the merge! I can go through tonight and update my network interface to be typed :100: I'm working about 30% on this right now as I clean up and add a few quick features to our app that just launched. But I pretty much always have slack open so a chat would be awesome!

**Comment by stubailo** [comment]
@jbaxleyiii can you join me on the MeteorChef slack? http://slack.themeteorchef.com/

**Comment by arlair** [comment]
I don't know if this is still helpful, but I believe you can use any JavaScript if you require it (not import as that does needs types).

const importEverything = require('myJavaScriptLibrary');
const { importThisOneThing } = require('myJavaScriptLibrary');

**Comment by stubailo** [comment]
Ooh, good to know! So far, I've been able to find type definitions online for every library we are using, but it will be helpful to know we don't absolutely need them!

**Comment by arlair** [comment]
I noticed you have typings reference comments shown in the code in the pull request. I believe that is also not required anymore with recent versions of TypeScript. Good luck.

**Comment by stubailo** [comment]
Yep, you'll see that the master branch doesn't have those anymore!

**Pull request merged**